coderd/prometheusmetrics/collector.go

```diff
@@ -0,0 +1,80 @@
+package prometheusmetrics
+
+import (
+	"sync"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type CachedGaugeVec struct {
+	m sync.Mutex
+
+	gaugeVec *prometheus.GaugeVec
+	records  []vectorRecord
+}
+
+var _ prometheus.Collector = new(CachedGaugeVec)
+
+type VectorOperation int
+
+const (
+	VectorOperationAdd VectorOperation = iota
+	VectorOperationSet
+)
+
+type vectorRecord struct {
+	operation   VectorOperation
+	value       float64
+	labelValues []string
+}
+
+func NewCachedGaugeVec(gaugeVec *prometheus.GaugeVec) *CachedGaugeVec {
+	return &CachedGaugeVec{
+		gaugeVec: gaugeVec,
+	}
+}
+
+func (v *CachedGaugeVec) Describe(desc chan<- *prometheus.Desc) {
+	v.m.Lock()
+	defer v.m.Unlock()
+
+	v.gaugeVec.Describe(desc)
+}
```

Comment:
This actually does not need the mutex. `Describe` is safe and does not return the counter, which is what you are protecting.

`Describe` is not really called much in prod, if ever, so it's not that big of a deal.

Reply:
I wasn't sure if `Describe` should be guarded or not, so thanks for raising it. I removed the mutex from the function.